packages/cozy-stack-client/src/xhrFetch.js

```diff
@@ -0,0 +1,55 @@
+const headersFromString = headerString => {
+  return new Headers(
+    headerString
+      .split('\r\n')
+      .map(x => x.split(':', 2))
+      .filter(x => x.length == 2)
+  )
+}
+
+/**
+ * Returns a `fetch()` like response but uses XHR.
+ * XMLHTTPRequest provides upload progress events unlike fetch.
+ *
+ * @param {string} fullpath - Route path
+ * @param {object} options - Fetch options
+ * @param {Function} options.onUploadProgress - Callback to receive upload progress events
+ */
+const fetchWithXMLHttpRequest = async (fullpath, options) => {
+  const response = await new Promise((resolve, reject) => {
+    const xhr = new XMLHttpRequest()
+
+    if (options.onUploadProgress) {
+      xhr.upload.addEventListener('progress', options.onUploadProgress, false)
+    }
+
+    xhr.onload = function() {
+      if (this.readyState == 4) {
+        resolve(this)
+      } else {
+        reject(this)
+      }
+    }
+
+    xhr.open(options.method, fullpath, true)
+    xhr.withCredentials = true
+    for (let [headerName, headerValue] of Object.entries(options.headers)) {
+      xhr.setRequestHeader(headerName, headerValue)
+    }
+
+    xhr.send(options.body)
+  })
+
+  return {
+    headers: headersFromString(response.getAllResponseHeaders()),
+    ok: response.status >= 200 && response.status < 300,
+    text: async () => response.responseText,
+    json: async () => JSON.parse(response.responseText)
+  }
+}
+
+const shouldXMLHTTPRequestBeUsed = (method, path, options) => {
+  return Boolean(options.onUploadProgress)
```

Comment:
It seems that there is a few interrogation marks on for the iOS supports here : https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Using_XMLHttpRequest

We should maybe be more strict than just check if there is this option?

Reply:
we use `xhr.onload`. And we have a "?" for safari ios support on "load event". Same thing for `progess`. 

> You mean I should add a user agent check ?

Nop, adding a test like : 
if(xhr.onload !== undefined && xhr.onprogress !== undefined)

to test if the method we used are available

Reply:
I added a navigator user agent check, and made the event binding more robust (by checking if xhr.upload exists).